### DISCUSSION TITLE
[Preview] Setting preview cookies doesn't bypass static generation

### DISCUSSION THREAD

**Discussion by justintemps** [vercel/next.js#19944] [Help]
I opened an issue for this as a bug #19714 , which I'm pretty sure is correct, but I'm sharing it here as well in case someone has ever encountered this situation before and found a work around.

tl;dr I can't get Preview Mode to work in production because `getStaticProps` never gets called even though the preview cookies are set. Here's a synopsis (I'm doing this with [Prismic CMS](https://prismic.io/))

**1.** From the Prismic Editor UI, the user clicks on the Preview button which redirects to 
```
https://myapp.prismic.io/previews/session/draft?previewId=X65pGRMAACEAEC0d&document=X7zhlRIAACIAz2iO&version=X85fixAAACQAOrMN`
```

**2.** Prismic then redirects to 
```
https://my.actual.app/api/preview?token=https%3A%2F%2Filo-voices.prismic.io%2Fpreviews%2FX8IC3RAAACQAA0pG%3AX85fixAAACQAOrMN%3FwebsitePreviewId%3DX65pGRMAACEAEC0d&documentId=X7zhlRIAACIAz2iO
```

**3.** The `api/preview` endpoint then gets the `redirectUrl` from Prismic and redirects to `/redirectUrl`, setting the Next.js preview cookies.

```
set-cookie: __prerender_bypass=391b9dd5be2db3f82d091836f424ccd3; Path=/; HttpOnly; Secure; SameSite=None

set-cookie: __next_preview_data=eyJhbGciOiJIUzI1NiIsInR5c...; Path=/; HttpOnly; Secure; SameSite=None
```

I am able to load the page in preview mode immediately after deploying and rebuilding the app to the server (Digital Ocean App Platform). However, if I leave Preview mode and load the page again, I am no longer able to view it in Preview Mode afterwards. Successive attempts redirect me to the correct page with the cookies set, but without `getStaticProps` getting called.

On localhost, bizarrely, it works fine on FireFox but not at all on Chrome.

Any help with this would be hugely appreciated!

**Comment by vladpereverzyev** [comment]
The behavior described here does look abnormal.

In Next.js, Preview Mode is specifically designed to bypass the statically generated version of a page. When the preview cookies are valid, the request should cause `getStaticProps()` to run again, with `context.preview === true` and the corresponding `context.previewData` available.

Since the issue only appears after leaving Preview Mode and then trying to enter it again, I would investigate the following areas:

1. **Next.js version**

   This discussion is from December 2020, when Preview Mode was still relatively new and had several edge cases. I would first verify the exact Next.js version being used and check whether the issue was already fixed in a later release.

2. **Use the official Preview Mode API**

   The `/api/preview` endpoint should use `res.setPreviewData(...)` rather than manually setting `__prerender_bypass` or `__next_preview_data`.

3. **Verify that `getStaticProps()` is actually being executed**

   I would temporarily add:

   ```js
   export async function getStaticProps(context) {
     console.log({
       preview: context.preview,
       previewData: context.previewData,
     })

     // ...
   }
   ```

   If the preview cookies are present but `getStaticProps()` isn't executed at all, the problem is likely related to caching, routing, or static generation rather than Prismic itself.

4. **Inspect the preview cookies**

   I would check the following in Chrome DevTools:

   * `__prerender_bypass`
   * `__next_preview_data`
   * Domain
   * Path
   * `Secure`
   * `SameSite`
   * Whether duplicate cookies exist for different domains/paths

   The fact that Firefox and Chrome behave differently is particularly interesting and makes the cookie handling worth investigating.

5. **Check the CDN/proxy/cache layer**

   Since the application is deployed on DigitalOcean, I would also verify whether there is any CDN, reverse proxy, or caching layer in front of Next.js.

   A cached response could prevent the request from reaching the Next.js rendering pipeline, effectively making it look as if Preview Mode is being ignored.

6. **Check how Preview Mode is disabled**

   The endpoint used to exit Preview Mode should use the appropriate Next.js API (`res.clearPreviewData()` for the versions of that era) rather than manually deleting or modifying the preview cookies.

The fact that rebuilding the application makes Preview Mode work again is particularly interesting. It suggests that the rebuild may be resetting some state related to static generation or caching, rather than actually fixing the Preview Mode cookies themselves.

My preferred way to debug this would be to completely isolate Prismic first:

* `/api/preview` → `res.setPreviewData()`
* redirect to `/test`
* `/test` → `getStaticProps()`
* log `context.preview`

If the problem can be reproduced with this minimal setup, then I would investigate Next.js and the deployment/cache layer.

If the minimal example works correctly, then I would focus on the Prismic integration, redirects, or caching behavior.

Also, because this GitHub discussion dates back to **December 2020**, I would be careful about applying its workaround directly to a modern Next.js application. The rendering and caching architecture of Next.js has changed significantly since then.
